ParabolInc/parabol
# feat: speed up repo integrations query

Fix https://github.com/ParabolInc/parabol/issues/6007

The `repoIntegrations` query is very expensive as it reaches out to all of our integrations even though the user probably wants a few of their most recently used repoIntegrations.

In this PR, I've removed the `allRepoIntegrations` query, the `useAllIntegrations` hook, and the query to our db in `repoIntegrationHelpers` which returned the previously used integrations.

Instead, I've implemented a cache in Redis so devs can just query `repoIntegrations`. 

The Redis cache consists of `allRepoIntegrations` and `prevUsedRepoIntegrations`:

- `allRepoIntegrations` is a key value pair with a 90-day TTL that gets updated every time we reach out to the network
- `prevUsedRepoIntegrations` is a sorted set which uses the most recently used unix timestamp as the score

Sorry this is such a large PR! Refactoring the way `repoIntegrations` work requires changing a bunch of different files. I also had to upgrade Redis to get the types I needed which meant changing several additional files. 


### To test

- [ ] Can push a task to GitHub, GitLab, Jira, Jira Server, and Azure DevOps
- [ ] Can create a new GitHub, GitLab, Jira, Jira Server, and Azure DevOps issue in Sprint Poker
- [ ] After pushing a task to an integration or creating an issue, refresh the page, open the task footer integration menu, and see that the most recently used `repoIntegration` is now at the top of the menu
- [ ] After pushing a few tasks to integrations, remove one of the integrations, e.g. GitHub, refresh the page, open the task footer menu again, and see that the menu no longer includes that integration
- [ ] Add a new integration, e.g. GitLab, refresh the page, and see that the integration shows up in the task footer menu
- [ ] In the task footer menu, search for a repoIntegration that doesn't exist. See that `repoIntegrations` reaches out to the APIs 

Ideally, we wouldn't need to refresh to see the changes, but that's the way it was previously working. As this PR is large enough, I considered changing that to be out of scope.

Creating a new issue in Azure results in an undefined project title, see issue: https://github.com/ParabolInc/parabol/issues/7266

https://www.loom.com/share/05f8a6d4147e415cac8cf732f233ad9f
